予約編集の保存エラー2

@claude
予約を編集した際の保存がエラーになります。
througly thinkして原因を徹底的に分析してください。

![Image](https://github.com/user-attachments/assets/1809fa57-6b64-400d-9cb0-d6689f6e93fe)
 Raw room data from database: Array(12)
 Raw room 0: Object
 Mapped room 0: Object
 Raw room 1: Object
 Mapped room 1: Object
 Raw room 2: Object
 Mapped room 2: Object
 Raw room 3: Object
 Mapped room 3: Object
 Raw room 4: Object
 Mapped room 4: Object
 Raw room 5: Object
 Mapped room 5: Object
 Raw room 6: Object
 Mapped room 6: Object
 Raw room 7: Object
 Mapped room 7: Object
 Raw room 8: Object
 Mapped room 8: Object
 Raw room 9: Object
 Mapped room 9: Object
 Raw room 10: Object
 Mapped room 10: Object
 Raw room 11: Object
 Mapped room 11: Object
 Mapped room data: Array(12)
 Raw room data from database: Array(12)
 Raw room 0: Object
 Mapped room 0: Object
 Raw room 1: Object
 Mapped room 1: Object
 Raw room 2: Object
 Mapped room 2: Object
 Raw room 3: Object
 Mapped room 3: Object
 Raw room 4: Object
 Mapped room 4: Object
 Raw room 5: Object
 Mapped room 5: Object
 Raw room 6: Object
 Mapped room 6: Object
page-a976e6bdcc9c72bc.js:1 Raw room 7: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 7: Object
page-a976e6bdcc9c72bc.js:1 Raw room 8: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 8: Object
page-a976e6bdcc9c72bc.js:1 Raw room 9: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 9: Object
page-a976e6bdcc9c72bc.js:1 Raw room 10: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 10: Object
page-a976e6bdcc9c72bc.js:1 Raw room 11: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 11: Object
page-a976e6bdcc9c72bc.js:1 Mapped room data: Array(12)
page-a976e6bdcc9c72bc.js:1 Raw room data from database: Array(12)
page-a976e6bdcc9c72bc.js:1 Raw room 0: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 0: Object
page-a976e6bdcc9c72bc.js:1 Raw room 1: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 1: Object
page-a976e6bdcc9c72bc.js:1 Raw room 2: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 2: Object
page-a976e6bdcc9c72bc.js:1 Raw room 3: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 3: Object
page-a976e6bdcc9c72bc.js:1 Raw room 4: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 4: Object
page-a976e6bdcc9c72bc.js:1 Raw room 5: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 5: Object
page-a976e6bdcc9c72bc.js:1 Raw room 6: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 6: Object
page-a976e6bdcc9c72bc.js:1 Raw room 7: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 7: Object
page-a976e6bdcc9c72bc.js:1 Raw room 8: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 8: Object
page-a976e6bdcc9c72bc.js:1 Raw room 9: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 9: Object
page-a976e6bdcc9c72bc.js:1 Raw room 10: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 10: Object
page-a976e6bdcc9c72bc.js:1 Raw room 11: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 11: Object
page-a976e6bdcc9c72bc.js:1 Mapped room data: Array(12)
page-a976e6bdcc9c72bc.js:1 Raw room data from database: Array(12)
page-a976e6bdcc9c72bc.js:1 Raw room 0: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 0: Object
page-a976e6bdcc9c72bc.js:1 Raw room 1: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 1: Object
page-a976e6bdcc9c72bc.js:1 Raw room 2: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 2: Object
page-a976e6bdcc9c72bc.js:1 Raw room 3: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 3: Object
page-a976e6bdcc9c72bc.js:1 Raw room 4: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 4: Object
page-a976e6bdcc9c72bc.js:1 Raw room 5: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 5: Object
page-a976e6bdcc9c72bc.js:1 Raw room 6: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 6: Object
page-a976e6bdcc9c72bc.js:1 Raw room 7: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 7: Object
page-a976e6bdcc9c72bc.js:1 Raw room 8: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 8: Object
page-a976e6bdcc9c72bc.js:1 Raw room 9: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 9: Object
page-a976e6bdcc9c72bc.js:1 Raw room 10: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 10: Object
page-a976e6bdcc9c72bc.js:1 Raw room 11: Object
page-a976e6bdcc9c72bc.js:1 Mapped room 11: Object
page-a976e6bdcc9c72bc.js:1 Mapped room data: Array(12)
/api/booking/41e225d1-08b7-4a3e-b337-c5e8728c3591:1 
            
            
           Failed to load resource: the server responded with a status of 500 ()
3090-d0a0282ccaa0aa7b.js:1 Error updating booking: Error: 予約の更新に失敗しました
    at M (page-8020c019d3b9abf3.js:1:10538)
window.console.error @ 3090-d0a0282ccaa0aa7b.js:1

＠claude
エラーが継続しています。
supabaseのSQLも含めて、現状の実装状況を徹底的に deeply and througly think してください。



/api/booking/e7cc87fd-5e7a-48c5-ae00-b8dcabdf3091:1 
            
            
           Failed to load resource: the server responded with a status of 500 ()
3090-d0a0282ccaa0aa7b.js:1 Error updating booking: Error: 予約の更新に失敗しました
    at M (page-957fcc3eeb8b1d2b.js:1:11120)
window.console.error @ 3090-d0a0282ccaa0aa7b.js:1


{"id":"e7cc87fd-5e7a-48c5-ae00-b8dcabdf3091","board_project_id":null,"status":"confirmed","start_date":"2025-09-17","end_date":"2025-11-12","nights":56,"pax_total":3,"pax_adults":0,"pax_adult_leaders":0,"pax_students":2,"pax_children":0,"pax_infants":1,"pax_babies":0,"guest_name":"aa","guest_email":"aa@test.com","guest_phone":"01210123456","guest_org":null,"purpose":null,"room_amount":560000,"pax_amount":1080792,"addon_amount":280000,"subtotal_amount":1920792,"total_amount":1920792,"notes":null,"created_at":"2025-06-06T15:35:54.909898+00:00","updated_at":"2025-06-06T15:35:54.909898+00:00","created_by":null,"project_rooms":[{"id":"61450024-800b-4c09-8ebd-38d7650c7fe0","rooms":{"name":"2年1組","floor":"2F","room_id":"R203","capacity":5,"amenities":["エアコン","照明"],"is_active":true,"room_rate":5000,"room_type":"small_c","created_at":"2025-05-30T11:39:00.539818+00:00","updated_at":"2025-05-30T11:39:00.539818+00:00","usage_type":"private","description":"5名定員の個室"},"amount":15680000,"nights":56,"room_id":"R203","room_rate":280000,"created_at":"2025-06-06T15:35:56.290663+00:00","project_id":"e7cc87fd-5e7a-48c5-ae00-b8dcabdf3091","assigned_pax":2},{"id":"7756eeee-7666-4c2e-82de-6fd02b9b6b9f","rooms":{"name":"3年3組","floor":"3F","room_id":"R303","capacity":5,"amenities":["エアコン","照明"],"is_active":true,"room_rate":5000,"room_type":"small_c","created_at":"2025-05-30T11:39:00.539818+00:00","updated_at":"2025-05-30T11:39:00.539818+00:00","usage_type":"private","description":"5名定員の個室"},"amount":15680000,"nights":56,"room_id":"R303","room_rate":280000,"created_at":"2025-06-06T15:35:56.290663+00:00","project_id":"e7cc87fd-5e7a-48c5-ae00-b8dcabdf3091","assigned_pax":2}]}

@claude
再度エラーがでています。

/api/booking/401bd338-a208-4c9c-9e9d-728b27df82c5:1 
            
            
           Failed to load resource: the server responded with a status of 500 ()
3090-d0a0282ccaa0aa7b.js:1 Error updating booking: Error: 予約の更新に失敗しました
    at M (page-6c5d32a77c1ca76d.js:1:11028)
window.console.error @ 3090-d0a0282ccaa0aa7b.js:1
page-6c5d32a77c1ca76d.js:1 
            
            
           PUT https://v0-simple-next-js-app-two.vercel.app/api/booking/401bd338-a208-4c9c-9e9d-728b27df82c5 500 (Internal Server Error)
M @ page-6c5d32a77c1ca76d.js:1
iG @ af67301d-8abbc52e6017d800.js:1
(anonymous) @ af67301d-8abbc52e6017d800.js:1
nS @ af67301d-8abbc52e6017d800.js:1
i4 @ af67301d-8abbc52e6017d800.js:1
ce @ af67301d-8abbc52e6017d800.js:1
s9 @ af67301d-8abbc52e6017d800.js:1
3090-d0a0282ccaa0aa7b.js:1 Error updating booking: Error: 予約の更新に失敗しました
    at M (page-6c5d32a77c1ca76d.js:1:11028)
window.console.error @ 3090-d0a0282ccaa0aa7b.js:1
M @ page-6c5d32a77c1ca76d.js:1
await in M
iG @ af67301d-8abbc52e6017d800.js:1
(anonymous) @ af67301d-8abbc52e6017d800.js:1
nS @ af67301d-8abbc52e6017d800.js:1
i4 @ af67301d-8abbc52e6017d800.js:1
ce @ af67301d-8abbc52e6017d800.js:1
s9 @ af67301d-8abbc52e6017d800.js:1

--- a/app/(dashboard)/booking/[id]/page.tsx
+++ b/app/(dashboard)/booking/[id]/page.tsx
@@ -63,6 +63,25 @@ export default function BookingDetailPage() {
     try {
       setIsSaving(true)
       
+      // Validate PAX constraint before sending to server
+      const paxTotal = editForm.pax_total || 0
+      const paxSum = (editForm.pax_adults || 0) + 
+                    (editForm.pax_adult_leaders || 0) + 
+                    (editForm.pax_students || 0) + 
+                    (editForm.pax_children || 0) + 
+                    (editForm.pax_infants || 0) + 
+                    (editForm.pax_babies || 0)
+      
+      if (paxTotal <= 0) {
+        alert('宿泊人数は1名以上である必要があります')
+        return
+      }
+      
+      if (paxTotal !== paxSum) {
+        alert(`人数内訳の合計(${paxSum}名)が宿泊人数(${paxTotal}名)と一致していません。内訳を確認してください。`)
+        return
+      }
+      
       // Clean the data to only include database-valid updatable fields
       const updateData: Record<string, any> = {}
       
@@ -278,12 +297,39 @@ export default function BookingDetailPage() {
                     <div className="space-y-2">
                       <Label>宿泊人数</Label>
                       {isEditing ? (
-                        <Input
-                          type="number"
-                          min="1"
-                          value={editForm.pax_total || 0}
-                          onChange={(e) => setEditForm({ ...editForm, pax_total: Number.parseInt(e.target.value) || 0 })}
-                        />
+                        <div className="space-y-2">
+                          <Input
+                            type="number"
+                            min="1"
+                            value={editForm.pax_total || 0}
+                            onChange={(e) => {
+                              const newTotal = Number.parseInt(e.target.value) || 0
+                              // When pax_total changes, adjust the breakdown to maintain constraint
+                              // Keep existing breakdown proportions but adjust to match new total
+                              const currentSum = (editForm.pax_adults || 0) + 
+                                                (editForm.pax_adult_leaders || 0) + 
+                                                (editForm.pax_students || 0) + 
+                                                (editForm.pax_children || 0) + 
+                                                (editForm.pax_infants || 0) + 
+                                                (editForm.pax_babies || 0)
+                              
+                              if (currentSum > 0 && newTotal !== currentSum) {
+                                // Simple adjustment: put the difference in students field
+                                const difference = newTotal - currentSum
+                                setEditForm({ 
+                                  ...editForm, 
+                                  pax_total: newTotal,
+                                  pax_students: Math.max(0, (editForm.pax_students || 0) + difference)
+                                })
+                              } else {
+                                setEditForm({ ...editForm, pax_total: newTotal })
+                              }
+                            }}
+                          />
+                          <div className="text-xs text-muted-foreground">
+                            内訳合計: {((editForm.pax_adults || 0) + (editForm.pax_adult_leaders || 0) + (editForm.pax_students || 0) + (editForm.pax_children || 0) + (editForm.pax_infants || 0) + (editForm.pax_babies || 0))}名
+                          </div>
+                        </div>
                       ) : (
                         <p className="text-sm">{booking.pax_total}名</p>
                       )}
@@ -310,6 +356,72 @@ export default function BookingDetailPage() {
                       )}
                     </div>
                   </div>
+
+                  {/* PAX Breakdown Details (only shown in edit mode) */}
+                  {isEditing && (
+                    <div className="space-y-2">
+                      <Label>人数内訳詳細</Label>
+                      <div className="grid grid-cols-2 gap-4 p-3 border rounded-lg bg-muted/20">
+                        <div className="space-y-2">
+                          <Label className="text-xs">大人</Label>
+                          <Input
+                            type="number"
+                            min="0"
+                            value={editForm.pax_adults || 0}
+                            onChange={(e) => setEditForm({ ...editForm, pax_adults: Number.parseInt(e.target.value) || 0 })}
+                          />
+                        </div>
+                        <div className="space-y-2">
+                          <Label className="text-xs">引率者</Label>
+                          <Input
+                            type="number"
+                            min="0"
+                            value={editForm.pax_adult_leaders || 0}
+                            onChange={(e) => setEditForm({ ...editForm, pax_adult_leaders: Number.parseInt(e.target.value) || 0 })}
+                          />
+                        </div>
+                        <div className="space-y-2">
+                          <Label className="text-xs">学生</Label>
+                          <Input
+                            type="number"
+                            min="0"
+                            value={editForm.pax_students || 0}
+                            onChange={(e) => setEditForm({ ...editForm, pax_students: Number.parseInt(e.target.value) || 0 })}
+                          />
+                        </div>
+                        <div className="space-y-2">
+                          <Label className="text-xs">子供</Label>
+                          <Input
+                            type="number"
+                            min="0"
+                            value={editForm.pax_children || 0}
+                            onChange={(e) => setEditForm({ ...editForm, pax_children: Number.parseInt(e.target.value) || 0 })}
+                          />
+                        </div>
+                        <div className="space-y-2">
+                          <Label className="text-xs">幼児</Label>
+                          <Input
+                            type="number"
+                            min="0"
+                            value={editForm.pax_infants || 0}
+                            onChange={(e) => setEditForm({ ...editForm, pax_infants: Number.parseInt(e.target.value) || 0 })}
+                          />
+                        </div>
+                        <div className="space-y-2">
+                          <Label className="text-xs">乳児</Label>
+                          <Input
+                            type="number"
+                            min="0"
+                            value={editForm.pax_babies || 0}
+                            onChange={(e) => setEditForm({ ...editForm, pax_babies: Number.parseInt(e.target.value) || 0 })}
+                          />
+                        </div>
+                      </div>
+                      <div className="text-xs text-muted-foreground">
+                        ※ 内訳合計が宿泊人数と一致する必要があります
+                      </div>
+                    </div>
+                  )}
 
                   <div className="space-y-2">
                     <Label>割り当て部屋</Label>

--- a/app/api/booking/[id]/route.ts
+++ b/app/api/booking/[id]/route.ts
@@ -74,6 +74,15 @@ export async function PUT(request: NextRequest, { params }: RouteContext) {
     console.log("- Booking ID:", id)
     console.log("- Request body keys:", Object.keys(body))
     console.log("- Request body:", JSON.stringify(body, null, 2))
+    console.log("- Request body PAX values:", {
+      pax_total: body.pax_total,
+      pax_adults: body.pax_adults,
+      pax_adult_leaders: body.pax_adult_leaders,
+      pax_students: body.pax_students,
+      pax_children: body.pax_children,
+      pax_infants: body.pax_infants,
+      pax_babies: body.pax_babies
+    })
 
     // プロジェクトの存在確認
     const { data: existingProject, error: fetchError } = await supabase
@@ -153,45 +162,74 @@ export async function PUT(request: NextRequest, { params }: RouteContext) {
     }
 
     console.log("BOOKING_EDIT_DEBUG - PAX validation fields:", paxFields)
+    console.log("BOOKING_EDIT_DEBUG - PAX field types:", {
+      pax_total: typeof paxFields.pax_total,
+      pax_adults: typeof paxFields.pax_adults,
+      pax_adult_leaders: typeof paxFields.pax_adult_leaders,
+      pax_students: typeof paxFields.pax_students,
+      pax_children: typeof paxFields.pax_children,
+      pax_infants: typeof paxFields.pax_infants,
+      pax_babies: typeof paxFields.pax_babies
+    })
 
-    // Calculate PAX sum for validation
-    const paxSum = (paxFields.pax_adults || 0) + 
-                   (paxFields.pax_adult_leaders || 0) + 
-                   (paxFields.pax_students || 0) + 
-                   (paxFields.pax_children || 0) + 
-                   (paxFields.pax_infants || 0) + 
-                   (paxFields.pax_babies || 0)
+    // Calculate PAX sum for validation - handle data types carefully
+    const paxSum = Number(paxFields.pax_adults || 0) + 
+                   Number(paxFields.pax_adult_leaders || 0) + 
+                   Number(paxFields.pax_students || 0) + 
+                   Number(paxFields.pax_children || 0) + 
+                   Number(paxFields.pax_infants || 0) + 
+                   Number(paxFields.pax_babies || 0)
+
+    const paxTotal = Number(paxFields.pax_total)
 
     console.log("BOOKING_EDIT_DEBUG - PAX calculation:")
-    console.log("- pax_total:", paxFields.pax_total)
-    console.log("- calculated sum:", paxSum)
-    console.log("- constraint validation:", paxFields.pax_total === paxSum)
+    console.log("- pax_total (raw):", paxFields.pax_total, typeof paxFields.pax_total)
+    console.log("- pax_total (converted):", paxTotal, typeof paxTotal)
+    console.log("- calculated sum:", paxSum, typeof paxSum)
+    console.log("- constraint validation (===):", paxTotal === paxSum)
+    console.log("- constraint validation (==):", paxTotal == paxSum)
+    console.log("- individual PAX values (converted):", {
+      adults: Number(paxFields.pax_adults || 0),
+      adult_leaders: Number(paxFields.pax_adult_leaders || 0),
+      students: Number(paxFields.pax_students || 0),
+      children: Number(paxFields.pax_children || 0),
+      infants: Number(paxFields.pax_infants || 0),
+      babies: Number(paxFields.pax_babies || 0)
+    })
 
     // Validate PAX constraint (must match database constraint)
-    if (paxFields.pax_total <= 0) {
+    if (paxTotal <= 0) {
       console.error("BOOKING_EDIT_ERROR - PAX constraint failed: pax_total must be > 0")
       return NextResponse.json(
         { 
           error: "PAX validation failed: Total guests must be greater than 0",
-          details: `pax_total: ${paxFields.pax_total}`
+          details: `pax_total: ${paxTotal}`
         },
         { status: 400 }
       )
     }
 
-    if (paxFields.pax_total !== paxSum) {
+    if (paxTotal !== paxSum) {
       console.error("BOOKING_EDIT_ERROR - PAX constraint failed: pax_total ≠ sum of breakdown")
       return NextResponse.json(
         { 
           error: "PAX validation failed: Total guests must equal sum of breakdown",
-          details: `pax_total: ${paxFields.pax_total}, breakdown sum: ${paxSum}`,
-          breakdown: paxFields
+          details: `pax_total: ${paxTotal}, breakdown sum: ${paxSum}`,
+          breakdown: paxFields,
+          convertedBreakdown: {
+            adults: Number(paxFields.pax_adults || 0),
+            adult_leaders: Number(paxFields.pax_adult_leaders || 0),
+            students: Number(paxFields.pax_students || 0),
+            children: Number(paxFields.pax_children || 0),
+            infants: Number(paxFields.pax_infants || 0),
+            babies: Number(paxFields.pax_babies || 0)
+          }
         },
         { status: 400 }
       )
     }
 
-    // Data type validation for numeric fields
+    // Data type validation and conversion for numeric fields
     const numericFields = ['room_amount', 'pax_amount', 'addon_amount', 'subtotal_amount', 'total_amount']
     for (const field of numericFields) {
       if (updateData[field as keyof ProjectUpdate] !== undefined) {
@@ -205,6 +243,30 @@ export async function PUT(request: NextRequest, { params }: RouteContext) {
           return NextResponse.json(
             { 
               error: `Invalid data type for ${field}`,
+              details: `Expected number, got ${typeof value}`
+            },
+            { status: 400 }
+          )
+        }
+      }
+    }
+
+    // Convert PAX fields to integers for database consistency
+    const paxFieldsToConvert = ['pax_total', 'pax_adults', 'pax_adult_leaders', 'pax_students', 'pax_children', 'pax_infants', 'pax_babies']
+    for (const field of paxFieldsToConvert) {
+      if (updateData[field as keyof ProjectUpdate] !== undefined) {
+        const value = updateData[field as keyof ProjectUpdate]
+        if (typeof value === 'string' && !isNaN(Number(value))) {
+          updateData[field as keyof ProjectUpdate] = parseInt(value, 10)
+          console.log(`BOOKING_EDIT_DEBUG - Converted PAX ${field} from string to integer: ${value}`)
+        } else if (typeof value === 'number') {
+          updateData[field as keyof ProjectUpdate] = Math.round(value)
+          console.log(`BOOKING_EDIT_DEBUG - Rounded PAX ${field} to integer: ${value}`)
+        } else if (value !== null && value !== undefined) {
+          console.error(`BOOKING_EDIT_ERROR - Invalid type for PAX ${field}: ${typeof value}`)
+          return NextResponse.json(
+            { 
+              error: `Invalid data type for PAX ${field}`,
               details: `Expected number, got ${typeof value}`
             },
             { status: 400 }

